src/odin/spare/spare_job.py

```diff
@@ -0,0 +1,383 @@
+
+    Path may start with s3:// or be a local filesystem file.
+    If the file doesn't exist, return an empty DataFrame with the columns from the schema.
+    """
+    columns = METADATA_COLUMNS | columns
+    # TODO assert that loaded file has same columns in same order
+    # TODO save the file timestamp to catch conflicts
+    if parquet_path.startswith("s3://"):
+        if s3.object_exists(parquet_path):
+            s3_stream = s3.stream_object(parquet_path)
+            return pl.read_parquet(s3_stream)
+        else:
+            return pl.DataFrame([], columns)
+    else:
+        try:
+            with open(parquet_path, "r") as f:
+                return pl.read_parquet(f)
+        except FileNotFoundError:
+            return pl.DataFrame([], columns)
+
+
+def write_parquet(df: pl.DataFrame, parquet_path: str, tmpdir: str) -> None:
+    """
+    Write a dataframe to a parquet file.
+
+    Path may start with s3:// or be a local filesystem file.
+
+    Logically, we only append to the parquet file,
+    but appending row groups is not well supported and is overkill for this use,
+    so we write the whole file.
+
+    TODO Prevent overwriting data if there's a conflict:
+    It's possible that the output parquet file was modified (perhaps by another instance of Odin)
+    after reading the file, while crunching the data. If so, we should not write the output file,
+    as that would overwrite those modifications
+    Instead, abort the job, leave the input files unarchived, and re-run the job immediately.
+    This way, the input files can be reprocessed and changes calculated based on the newest version
+    of the parquet file.
+    Implementation notes: For AWS, Save the ETag on read and pass it to If-Match on write.
+    For local files, use os.path.getmtime.
+    """
+    if parquet_path.startswith("s3://"):
+        # df.write_parquet() can't write to s3 directly, so write to local file, then upload it
+        tmp_path = os.path.join(tmpdir, parquet_path[5:])
+        os.makedirs(os.path.dirname(tmp_path), exist_ok=True)
+        df.write_parquet(tmp_path)
+        s3.upload_file(tmp_path, parquet_path)
+    else:
+        os.makedirs(os.path.dirname(parquet_path), exist_ok=True)
+        df.write_parquet(parquet_path)
+
+
+def datetime_from_delta_path_ms(path: str) -> int:
+    """
+    Extract a datetime from the filenames generated by Delta
+
+    Returns number of ms since the epoch.
+    Delta makes filenames like
+    2025/04/01/2025-04-01T18:12:33Z_https_example.com_path.gz
+    This extracts extracts and parses the ISO8601 timestamp
+    It ignores the yyyy/mm/dd/ prefix so it works if the file has been copied elsewhere.
+    """
+    iso8601 = path.split("/")[-1].split("_")[0]
+    if not iso8601.endswith("Z"):
+        raise Exception(
+            f'Expected iso8601 timestamp with a Z. Got "{iso8601}". From path "{path}".'
+        )
+    return int(isoparse(iso8601).timestamp() * 1000)
+
+
+def merge_new_data(
```

Comment:
I think this operation expects a parquet file that will always be able to fit into the system memory of the ECS.

How large do you expect this parquet file to get? Depending on the number of expected rows being added per day, the file could become too large to materialize in memory.